sources/packages/backend/apps/workers/src/services/disbursement-schedule/disbursement-schedule.service.ts

```diff
@@ -0,0 +1,179 @@
+          offeringIntensity,
+        );
+
+      let hasValidMSFAANumber = false;
+      if (previousSignedDisbursement) {
+        // checks if the MSFAA number is still valid.
+        // If the study period end date of the previously signed MSFAA is less than 2 years
+        // when compared to current study period start date, then MSFAA is considered to be valid.
+        hasValidMSFAANumber = this.msfaaNumberService.isMSFAANumberValid(
+          // Previously signed and completed application offering end date in considered the start date.
+          new Date(
+            previousSignedDisbursement.studentAssessment.offering.studyEndDate,
+          ),
+          // Start date of the offering of the current application is considered the end date.
+          new Date(firstDisbursement.studentAssessment.offering.studyStartDate),
+        );
+      }
+
+      if (hasValidMSFAANumber) {
+        // Reuse the MSFAA number.
+        msfaaNumberId = previousSignedDisbursement.msfaaNumber.id;
+      } else {
+        // Create a new MSFAA number in case the previous one is no longer valid.
+        const newMSFAANumber = await this.msfaaNumberService.createMSFAANumber(
+          studentId,
+          applicationId,
+          offeringIntensity,
+        );
+        msfaaNumberId = newMSFAANumber.id;
+      }
+    }
+    const msfaaNumber = { id: msfaaNumberId } as MSFAANumber;
+    firstDisbursement.msfaaNumber = msfaaNumber;
+    const disbursementsToUpdate = [firstDisbursement];
+
+    // If the assessment has second disbursement, associate MSFAA number
+    // and add to the disbursements to be updated.
+    if (secondDisbursement) {
+      secondDisbursement.msfaaNumber = msfaaNumber;
+      disbursementsToUpdate.push(secondDisbursement);
+    }
+
+    this.repo.save(disbursementsToUpdate);
+  }
+
+  /**
+   * Get the previously signed MSFAA for the student
+   * for given offering intensity if exist.
+   ** MSFAA is generated individually for full-time/part-time
+   ** offerings.
+   * @param studentId student.
+   * @param offeringIntensity  offering intensity.
+   * @returns previous signed MSFAA if exist.
+   */
+  async getPreviouslySignedDisbursement(
+    studentId: number,
+    offeringIntensity: OfferingIntensity,
+  ): Promise<DisbursementSchedule> {
+    return this.repo.findOne({
+      select: {
+        id: true,
+        studentAssessment: {
+          id: true,
+          offering: { id: true, studyEndDate: true },
+        },
+        msfaaNumber: { id: true },
+      },
+      relations: {
+        studentAssessment: { offering: true },
+        msfaaNumber: true,
+      },
+      where: {
+        studentAssessment: {
+          offering: { offeringIntensity },
+          application: {
+            student: { id: studentId },
+            applicationStatus: ApplicationStatus.Completed,
+          },
+        },
+        msfaaNumber: {
+          dateSigned: Not(IsNull()),
+          cancelledDate: IsNull(),
```

Comment:
Thanks for doing the conversion to the object query and adding the missing `cancelledDate` ❤️